File: 347-top-k-frequent-elements/top-k-frequent-elements.py
class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        freqs = {}
        for n in nums:
            freqs[n] = freqs.get(n, 0) + 1
        
        order = [[] for _ in range(len(nums))]

        for n in freqs:
            f = freqs[n]
            order[f - 1].append(n)
        
        res = []
        while k:
            l = order.pop()
            if not l:
                continue
            res.append(l.pop())
            k -= 1
            if l:
                order.append(l)
        return res

File: 347-top-k-frequent-elements/top-k-frequent-elements.java
import java.util.PriorityQueue;
import java.util.HashMap;

class Solution {
    public int[] topKFrequent(int[] nums, int k) {
        HashMap<Integer, Integer> f = new HashMap<>();
        for (int n : nums) {
            if (!f.containsKey(n)) {
                f.put(n, 0);
            }
            f.put(n,f.get(n) + 1);
        }

        PriorityQueue<Integer> pq = new PriorityQueue<>((a,b) -> f.get(b) - f.get(a));
        for (int key: f.keySet()) {
            pq.add(key);
        }

        int[] res = new int[k];
        for (int i = 0; i < k; i++) {
            res[i] = pq.poll();
        }

        return res;
    }
}